docs/SPIRVVersionsAndExtensionsHandling.rst

```diff
@@ -0,0 +1,250 @@
+translator.
+
+If ``--spirv-ext`` contains name of extension which is not know for the
+translator, it will emit error.
+
+Examples:
+
+* ``--spirv-ext=+SPV_KHR_no_integer_wrap_decoration,+SPV_INTEL_subgroups``
+* ``--spirv-ext=+all,-SPV_INTEL_fpga_loop_controls``
+
+.. warning::
+   Extension name cannot be allowed and disallowed at the same time: for inputs
+   like ``--spirv-ext=+SPV_INTEL_subgroups,-SPV_INTEL_subgroups`` translator
+   will emit error about invalid arguments.
+
+.. note::
+   Since by default during SPIR-V generation all extensions are disabled, this
+   means that ``-all,`` is implicitly added at the beggining of the
+   ``-spirv-ext`` value.
+
+If the translator encounters something that cannot be represented by set of
+allowed SPIR-V extensions (which might be empty), it does one of the following
+things:
+
+* ignores LLVM IR entity in the input file.
+
+  For example, ``nsw``/``nuw`` LLVM IR attributes can be ignored if it is not
+  allowed to generate SPIR-V 1.4 and ``SPV_KHR_no_integer_wrap_decoration``
+  extension is disallowed.
+
+* tries to represent LLVM IR entity with allowed instructions.
+
+  Translator could translate calls to a new built-in functions defined by some
+  extensions as usual call instructions without using special SPIR-V
+  instructions.
+
+  However, this could result in a strange SPIR-V and most likely will lead to
+  errors during consumption. Having that, translator should emit errors if it
+  encounters a call to a built-in function from an extension which must be
+  represented as a special SPIR-V instruction from extension which wasn't
+  allowed to be used. I.e. if translator knows that this certain LLVM IR entity
+  belongs to an extension functionality and this extension is disallowed, it
+  should emit error rather than emulating it.
+
+  This way of emulating missing instructions still might be taken by the
+  translator if it encounters something that can be represented in SPIR-V more
+  easily in presence of some extension, but can be legally emulated without it.
+  I.e. in situations similar to ``OpPtrEqual`` in SPIR-V less than 1.4
+
+  **TODO:** add example if such exists.
+
+* emits error if LLVM IR entity cannot be ignored and cannot be emulated using
+  available instructions.
+
+  For example, new built-in types defined by
+  ``cl_intel_device_side_avc_motion_estimation`` cannot be represented in SPIR-V
+  if ``SPV_INTEL_device_side_avc_motion_estimation`` is disallowed.
+
+SPIR-V Consumption step
+-----------------------
+
+By default, translator consumes SPIR-V regardless of list extensions which are
+used by the input file, i.e. all extensions are allowed by default during
+consumption step.
+
+.. note::
+   This is opposite to the generation step and this is done on purpose: to not
+   broke workflows of existing users of the translator.
+
+.. note::
+   Since by default during SPIR-V consumption all extensions are disabled, this
```

Comment:
> all extensions are **disabled**

Looks like it should be *enabled*, right ?

Reply:
Yes, indeed. Fixed